[Concurrency] Async Let (#1338)
apple/swift-evolution: proposals/mmmm-spawn-let.md

```swift
func makeDinner() async -> Meal {
  // Create a task group to scope the lifetime of our three child tasks
  return try await withThrowingTaskGroup(of: CookingTask.self) { group in
    // spawn three cooking tasks and execute them in parallel:
    group.spawn {
      CookingTask.veggies(try await chopVegetables())
    }
    group.spawn {
      CookingTask.meat(await marinateMeat())
    }
    group.spawn {
      CookingTask.oven(await preheatOven(temperature: 350))
    }

    // prepare variables to collect the results
    var veggies: [Vegetable]? = nil
    var meat: Meat? = nil
    var oven: Oven? = nil

    // collect the results
    for try await task in group {
      switch task {
      case .veggies(let v):
        veggies = v
      case .meat(let m):
        meat = m
      case .oven(let o):
        oven = o
      }
    }

    // ensure every variable was initialized as expected
    assert(veggies != nil)
    assert(meat != nil)

>     assert(oven != nil)

Author: The example is meant to be explicit about what is going on; unwraps make it look nicer but that's not the point here

    // prepare the ingredients
    var ingredients: [Ingredient] = veggies!
    ingredients.append(meat!)

    // and, finally, cook the meal, awaiting inside the group
    let dish = Dish(ingredients: ingredients)
    return try await oven!.cook(dish, duration: .hours(3))
  }
}
```

> The `withTaskGroup` scope explicitly delineates any potential concurrency, because it guarantees that any child tasks spawned within it are awaited on as the group scope exits. Any results can be collected using iterating through the group. Errors and cancellation are handled automatically for us by the group.

Reviewer: ```suggestion
The `withThrowingTaskGroup` scope explicitly delineates any potential concurrency, because it guarantees that any child tasks spawned within it are awaited on as the group scope exits. Any results can be collected by iterating through the group. Errors and cancellation are handled automatically for us by the group.
```

However, this example showcases the weaknesses of the TaskGroups very well: heterogenous result processing and variable initialization become very boiler plate heavy. While there exist ideas to make this boiler plate go away in future releases, with smarter analysis and type checking, the fundamental issue remains. 

> If we step back a little, we can notice that in the example each child task is really producing a *single value* and returning it back to the *parent task*, which then needs to assamble the pieces and proceed with calling some other function. We achieve this by preparing, and assigning into `Optional` variables dedicated for each of the spawned tasks. This is not ideal, since

Reviewer: `assamble` → `assemble`
```suggestion
If we step back a little, we can notice that in the example each child task is really producing a *single value* and returning it back to the *parent task*, which then needs to assemble the pieces and proceed with calling some other function. We achieve this by preparing, and assigning into `Optional` variables dedicated for each of the spawned tasks. This is not ideal, since
```

although the code is correct as written, modifying this code to add a variable is not only boilerplate heavy, but also potentially quite error prone, leading to runtime crashes due to the force-unwraps which a well written Swift program usually would not have to resort to. 

This dataflow pattern from child tasks to parents is very common, and we want to make it as lightweight and safe as possible.

## Proposed solution

> This proposal introduces a simple way to create child tasks with and await their results: `spawn let` declarations.

Reviewer: ```suggestion
This proposal introduces a simple way to create child tasks and await their results: `spawn let` declarations.
```

Using `spawn let`, our example looks like this:

```swift
// given: 
//   func chopVegetables() async throws -> [Vegetables]
//   func marinateMeat() async -> Meat

> //   func preheatOven(temperature: Int) -> Oven

Reviewer: ```suggestion
//   func preheatOven(temperature: Int) async -> Oven
```

  let dish = Dish(ingredients: await [try veggies, meat])
  return try await oven.cook(dish, duration: .hours(3))
}
```

`spawn let` is similar to a `let`, in that it defines a local constant that is initialized by the expression on the right-hand side of the `=`. However, it differs in that the initializer expression is evaluated in a separate, concurrently-executing child task. 

The child task begins running as soon as the `spawn let` is encountered. By default, child tasks use the global, width-limited,  concurrent executor, in the same manner as task group child-tasks do. It is a future direction to allow customizing which executor these should be executing on. On normal completion, the child task will initialize the variables in the `spawn let`.

The right-hand side of a `spawn let` expression can be thought of as an implicit `@Sendable closure`, similar to how the `detach { ... }` API works, however the resulting task is a *child task* of the currently executing task. Because of this, and the need to suspend to await the results of such expression, `spawn let` declarations may only occur within an asynchronous context, i.e. an `async` function or closure.

> For single statement expressions in the `spawn let` initializer, the `await` and `try` keywords may be omitted. The effects they represent carry through to the introduced constant and will have to be used on when waiting on the constant. For example, in the example shown above, the veggies are declared as `spawn let veggies = chopVegetables()`, and even through `chopVegetables` is `async` and `throws`, the `await` and `try` keywords do not have to be used on that line of code. Once waiting on the value of that `async let` constant, the compiler will enforce that the expression where the `veggies` appear must be covered by both `await` and some form of `try`.

Reviewer: `async let` → `spawn let`

```suggestion
For single statement expressions in the `spawn let` initializer, the `await` and `try` keywords may be omitted. The effects they represent carry through to the introduced constant and will have to be used when waiting on the constant. In the example shown above, the veggies are declared as `spawn let veggies = chopVegetables()`, and even through `chopVegetables` is `async` and `throws`, the `await` and `try` keywords do not have to be used on that line of code. Once waiting on the value of that `spawn let` constant, the compiler will enforce that the expression where the `veggies` appear must be covered by both `await` and some form of `try`.
```

> Because the main body of the function executes concurrently with its child tasks, it is possible that the parent task (the body of `makeDinner` in this example) will reach the point where it needs the value of an `spawn let` (say,`veggies`) before that value has been produced. To account for that, reading a variable defined by an `spawn let` is treated as a potential suspension point,

Reviewer: ```suggestion
Because the main body of the function executes concurrently with its child tasks, it is possible that the parent task (the body of `makeDinner` in this example) will reach the point where it needs the value of a `spawn let` (say,`veggies`) before that value has been produced. To account for that, reading a variable defined by a `spawn let` is treated as a potential suspension point,
```

  spawn let x = ... // error: 'spawn let' in a function that does not support concurrency
}

func syncMe(later: () -> String) { ... }
syncMe {
  spawn let x = ... // error: invalid conversion from 'async' function of type '() async -> String' to synchronous function type '() -> String'
}
```

A `spawn let` creates a child-task, which inherits its parent task's priority as well as task-local values. Semantically, this is equivalent to creating a one-off `TaskGroup` which spawns a single task and returns its result, however the implementation of `async let`s can make more assumptions and optimizations around the lifetime and usage of those values.

The child-task created to initialize the `spawn let` by default runs on the global concurrent, width-limited, executor that comes with the Swift Concurrency runtime. 

> Customizing the execution context of spawn lets is a future direction we are likely to explore with the introduction of Custom Executors.

The initializer of the `spawn let` can be thought of as a closure that runs the code contained within it in a separate task, very much like the explicit `group.spawn { <work here/> }` API of task groups.

> Similarily to the `group.spawn()` function, the closure is `@Sendable` and `nonisolated`, meaning that it cannot access non sendable state of the enclosing context. For example, it will result in a compile time-error, preventing a potential race condition, for a `spawn let` initializer to attempt mutating a closed over variable:

Author: Thank you, that word is my arch-nemesis :-)

```swift
spawn var x = nope() // error: 'spawn' can only be used with 'let' declarations
```

Other than having to be awaited to access its value, a `spawn let` behaves just like a typical `let`, as such it is not possible to pass it `inout` to other functions - simply because it is a `let`, and those may not be passed as `inout`.

#### Declaring `spawn let` with patterns

It is possible to create a `spawn let` where the left hand side is a pattern, e.g. a tuple, like this:

```swift
func left() async -> String { "l" }
func right() async -> String { "r" }

spawn let (l, r) = (left(), right())

await l // at this point `r` is also assumed awaited-on
```

To understand the execution semantics of the above snippet, we can remember the sugaring rule that the right-hand side of a `spawn let` effectively is just a concurrently executing asynchronous closure:

```swift
spawn let (l, r) = {
  return await (left(), right())
  // -> 

>   // return (left await(), await right())

Reviewer: ```suggestion
  // return (await left(), await right())
```

}
```

meaning that the entire initializer of the `spawn let` is a single task, and if multiple asynchronous function calls are made inside it, they are performed one-by one. This is a specific application of the general rule of `spawn let` initializers being allowed to omit a single leading `await` keyword before their expressions. Because in this example, we invoke two asynchronous functions to form a tuple, the await can be moved outside of the expression, and that await is what is omitted in the short hand form of the `spawn let` that we've seen in the first snippet.

This also means that as soon as we enter continue past the line of `await l` it is known that the `r` value also has completed sucessfully (and will not need to emit an "implicit await" which we'll discuss in detail below).

Another implication of these semantics is that if _any_ piece of the initializer throws, any await on such pattern declared `spawn let` shall be considered throwing, as they are initialized "together". To visualize this, let us consider the following:

```swift

> async let (yay, nay) = ("yay", throw Boom())

Reviewer: ```suggestion
spawn let (yay, nay) = ("yay", throw Boom())
```

try await yay // because the (yay, nay) initializer is throwing
```

Because we know that the right-hand side is simply a single closure, performing the entire initialization, we know that if any of the operations on the right hand size is throwing, the entire initializer will be considered throwing. As such, awaiting even the `yay` here must be ready for that initializer to have thrown and therefore must include the `try` keyword in addition to `await`.

### Awaiting `spawn let` values

> Since `spawn let`s introduce constants that will be "filled in later" by their right-hand-side concurrently-executing task, refering to them must be covered by an `await` keyword:

Reviewer: `refering` → `referring`

```suggestion
Since `spawn let`s introduce constants that will be "filled in later" by their right-hand-side concurrently-executing task, referring to them must be covered by an `await` keyword:
```

```swift
spawn let name = getName() 
spawn let surname = getSurname() 
await name
await surname
```

It is also possible to simply cover the entire expression where a `spawn let` is used with just a single `await`, similar to how the same can be done with `try`:

```swift
greet(await name, await surname)
await greet(name, surname)
// or even
await print("\(name) \(surname)")
```

If the initializer of the specific `spawn let` was throwing, then awaiting on the `spawn let` constant must be covered using a variant of the `try` keyword:

```swift
spawn let ohNo = throwThings()
try await ohNo
try? await ohNo
try! await ohNo
```

> Currently, it is required to cover every reference to a `spawn let` using the apropriate try and await keywords, like this:

Reviewer: `apropriate` → `appropriate`

```suggestion
Currently, it is required to cover every reference to a `spawn let` using the appropriate `try` and `await` keywords, like this:
```

```swift
spawn let yes = ""
spawn let ohNo = throwThings()

_ = await yes
_ = await yes
_ = try await ohNo
_ = try await ohNo
```

This is a simple rule and allows us to bring the feature forward already. It might be possible to employ control flow based analysis to enable "only the first reference to the specific `spawn let` on each control flow path has to be an `await`", as technically speaking, every following await will be a no-op and will not suspend as the value is already completed and the placeholder has been filled in.

### Implicit `spawn let` awaiting 

A `spawn let` that was declared but never awaited on *explicitly* as the scope in which it was declared exits, will be awaited on implicitly. These semantics are put in place to uphold the Structured Concurrency guarantees provided by `spawn let`.

To showcase these semantics, let us have a look at this function which spawns two child tasks, `fast` and `slow` but does not await on any of them:

```swift
func go() async { 
  spawn let f = fast() // 300ms
  spawn let s = slow() // 3seconds
  return "nevermind..."
  // implicitly: cancels f
  // implicitly: cancels s
  // implicitly: await f
  // implicitly: await s
}
```

> Assuming the execution times of fast and slow are as the comments next to them explain, the `go()` function will _always_ take at least 3 seconds to execute. Or to state the rule more generally, any structured invocation, will take as much time to return as much the longest of it child tasks takes to complete.

Reviewer: ```suggestion
Assuming the execution times of `fast()` and `slow()` are as the comments next to them explain, the `go()` function will _always_ take at least 3 seconds to execute. Or to state the rule more generally, any structured invocation will take as much time to return as the longest of its child tasks takes to complete.
```

As we return from the `go()` function without ever having awaited on the `f` or `s` values, both of them will be implicitly cancelled and awaited on before returning from the function `go()`. This is the very nature of structured concurrency, and avoiding this can _only_ be done by creating non-child tasks, e.g. by using `detach` or other future APIs which would allow creation of non-child tasks.

If we instead awaited on one of the values, e.g. the fast one (`f`) the emitted code would not need to implicitly cancel or await it, as this was already taken care of explicitly:

```swift
func go2() async {
  spawn let f = fast()
  spawn let s = slow()
  _ = await f
  return "nevermind..."
  // implicitly: cancels s
  // implicitly: awaits s
}
```

The duration of the `go2()` call remains the same, it is always `time(go2) == max(time(f), time(s))`.

> Special attention needs to be given to the `spawn let _ = ...` form of spawn let declarations. This form of is interesting because it creates a child-task of the right hand-side initializer, however actively chooses to ignore the result. Such declaration, and the associated child-task will run, and be awaited-on implicitly as the scope it was declared is about to exit - the same way as an un-used spawn let declaration would be.

Reviewer: ```suggestion
Special attention needs to be given to the `spawn let _ = ...` form of declarations. This form is interesting because it creates a child-task of the right-hand-side initializer, however it actively chooses to ignore the result. Such a declaration (and the associated child-task) will run and be awaited-on implicitly, as the scope it was declared in is about to exit — the same way as an unused `spawn let` declaration would be.
```

spawn let name = "Alice"
await greet { await name }
```

Notice how we are required to write the `await` inside the closure as well as in front of the `greet` function. This is on purpose as we do want to be explicit about the await inside the closure. 

The same applies to auto closures, in order to make it explicit that the await is happening _inside_ the closure rather than before it, it is required to await explicitly in parameter position where the auto closure is formed for the argument:

```swift
func greet(_ f: @autoclosure () async -> String) async -> String { await f() }

spawn let name = "Bob"
await greet(await name) // await on name is required, because autoclosure
```

It is *not* legal to escape a `spawn let` value to an escaping closure. This is because structures backing the spawn let implementation may be allocated on the stack rather than the heap. This makes them very efficient, and makes great use of the structured guarantees they have to adhere to. These optimizations, however, make it unsafe to pass them to any escaping contexts:

```swift
func greet(_ f: @escaping () async -> String) async -> String { somewhere = f; somewhere() }

spawn let name = "Bob"

> await greet(await name) // error: cannot escape 'spawn let' value

Reviewer: ```suggestion
await greet { await name } // error: cannot escape 'spawn let' value
```

```

> > Note: If Swift had a `@useImmediately` annotation that could be used together with even escaping closures, as they would "promise" to be called immediately without detaching or storing the closure elsewhere.

Reviewer 2: I'm not sure that this is true, because `detach` itself could be considered to `@useImmediately` its closure argument. To avoid lifetime-extending the `async let` task, any potential awaits of the value have to happen before it goes out of scope, which seems to me like it really does require the closure be nonescaping. We could however potentially allow for escaping closures to capture the value by awaiting it, with syntax like `{[await name] in name}`, perhaps.

Author: Ah I think you're right on that one, we cannot indefinitely extend the lifetime of the let after all... Thanks Joe.

I'll remove the note and add a future direction about capturing with the proposed `[await x]` capture list 👍

### `spawn let` error propagation

> While it is legal to declare a `spawn let` and never explicitly `await` on it, it also implies that we do not particuilary care about its result.

Reviewer: `particuilary` → `particularly`

```suggestion
While it is legal to declare a `spawn let` and never explicitly `await` on it, it also implies that we do not particularly care about its result.
```

This is the same as spawning a number of child-tasks in a task group, and not collecting their results, like so:

```swift
try await withThrowingTaskGroup(of: Int.self) { group in 
  group.spawn { throw Boom() }
                             
  return 0 // we didn't care about the child-task at all(!)
} // returns 0
```

The above TaskGroup example will ignore the `Boom` thrown by its child task. However, it _will_ await for the task (and any other tasks it had spawned) to run to completion before the `withThrowingTaskGroup` returns. If we wanted to surface all potential throws of tasks spawned in the group, we should have written: `for try await _ in group {}` which would have re-thrown the `Boom()`.

The same concept carries over to `spawn let`, where the scope of the group is replaced by the syntactic scope in which the `spawn let` was declared. For example, the following snippet is semantically equivalent to the above TaskGroup one:

```swift
// func boom() throws -> Int { throw Boom() }

func work() async -> Int {
  spawn let work: Int = boom()
  // never await work...
  return 0
  // implicitly: cancels work
  // implicitly: awaits work, discards errors
}
```

> This work function, will never throw, because we didn't await on the throwing `spawn let`. If we modified it to explicitly await on it, the compiler would force us to spell out not only the `await` but also the `try` keyword. The presence of the `try` keyword woult then force us to annotate the `work()` function as `throws`, as expected from normal, non-asynchronous code in Swift:

Reviewer: `woult` → `would`

```suggestion
This `work()` function will never throw, because we didn't await on the throwing `spawn let`. If we modified it to explicitly await on it, the compiler would force us to spell out not only the `await` but also the `try` keyword. The presence of the `try` keyword would then force us to annotate the `work()` function as `throws`, as expected from normal, non-asynchronous code in Swift:
```

      // concurrently, however without any real parallelism.
      spawn let x = process(first)
      spawn let y = process(second)
      // x and y do NOT execute in parallel
    }
  }
}
```

The details of the API remain to be seen, but the general ability to specify an executor for child-tasks is useful and will be considered in the future.

## Alternatives considered

### Alternative name: `async let`

The feature was previously known as `async let` yet as the work on Swift concurrency continued, based on community feedback as well as our own findings with regards of implied meanings of specific words used throughout all of Swift's concurrency APIs, we found that the word `async` is not quite right for this feature.

As `spawn let` should be thought of a specialization of TaskGroups and the child-tasks spawned by either of them inherit the same kind of semantics from their enclosing parent task, it makes sense to share the same name across the two features.

We feel that using the word `spawn` in _every_ case that involves the creation of a child-task is simple to understand and allows developers to learn and assume about both features in tandem. 

### Explicit futures

> As discussed in the [structured concurrency proposal](nnnn-structured-concurrency.md#Prominent-futures), we choose not to expose futures or `Task.Handle`s for child tasks in task groups, because doing so either can undermine the hierarchy of tasks, by escaping from their parent task group and being awaited on indefinitely later, or would result in there being two kinds of future, one of which dynamically asserts that it's only used within the task's scope. `spawn let` allows for future-like data flow from child tasks to parent, without the need for general-purpose futures to be exposed.

Reviewer: ```suggestion
As discussed in the [structured concurrency proposal](0304-structured-concurrency.md#prominent-futures), we choose not to expose futures or `Task.Handle`s for child tasks in task groups, because doing so either can undermine the hierarchy of tasks, by escaping from their parent task group and being awaited on indefinitely later, or would result in there being two kinds of future, one of which dynamically asserts that it's only used within the task's scope. `spawn let` allows for future-like data flow from child tasks to parent, without the need for general-purpose futures to be exposed.
```

### "Don't spawn tasks when in cancelled parent"

> It would be very confusing to have automatically "not run" if the parent task were cancelled. Such semantics are offered by task groups via the `group.spawnUnlessCancelled` API, however would be quite difficult to express using plain `let` declarations, as effectively all such declarations would have to become implicitly throwing, which would sacrifice their general usability. We are convienced that following through with the co-operative cancellation strategy works well for `spawn let` tasks, because it composes well with what all asynchronous functions should be handling cancellation to begin with: only when they want to, in apropriate places within their execution, and deciding by themselfes if they prefer to throw a `Task.CancellationError` or rather return a partial result when cancellation ocurrs.

Reviewer: `convienced` → `convinced`
`apropriate` → `appropriate`
`themselfes` → `themselves`
`ocurrs` → `occurs`

```suggestion
It would be very confusing to have `spawn let` tasks automatically "not run" if the parent task were cancelled. Such semantics are offered by task groups via the `group.spawnUnlessCancelled` API, however would be quite difficult to express using plain `let` declarations, as effectively all such declarations would have to become implicitly throwing, which would sacrifice their general usability. We are convinced that following through with the co-operative cancellation strategy works well for `spawn let` tasks, because it composes well with how all asynchronous functions should be handling cancellation to begin with: only when they want to, in appropriate places within their execution, and deciding by themselves if they prefer to throw a `Task.CancellationError` or rather return a partial result when cancellation occurs.
```

## Revision history

After initial pitch (as part of Structured Concurrency):

> - renamed `spawn let` to `spawn let` to be consistent with `spawn` usage in the rest of structured concurrency APIs,

Reviewer: ```suggestion
- renamed `async let` to `spawn let` to be consistent with `spawn` usage in the rest of structured concurrency APIs,
```

```swift
// very silly example to show limitations of `spawn let` when facing dynamic numbers of tasks
func toyParallelMapExactly2<A, B>(_ items: [A], f: (A) async -> B) async -> [B] { 
  assert(items.count == 2)
  spawn let f0 = f(items[0])
  spawn let f1 = f(items[1])
  
  return await [f0, f1]
}
```

And while the second example reads very nicely, it cannot work in practice to implement such parallel map function, because the size of the input `items` is not known (and we'd have to implement `1...n` versions of such function).

Another API which is not implementable with `spawn let` and will require using a task group is anything that requires some notion of completion order. Because `spawn let` declarations must be awaited on it is not possible to express "whichever completes first" and a task group must be used to implement such API. 

For example, the `race(left:right:)` function shown below, runs two child tasks in parallel, and returns whichever completed first. Such API is not possible to implement using spawn let and must be implemented using a group:

```swift
func race(left: () async -> Int, right: () async -> Int) async -> Int {

>   await withTaskGroup(of: Int) { 

Reviewer: ```suggestion
  await withTaskGroup(of: Int.self) {
```

    group.spawn { left() }
    group.spawn { right() }

    return await group.next()! // !-safe, there is at-least one result to collect
  }
}
```

#### Comparing with Task.Handle, and (not proposed) futures

It is worth comparing `spawn let` declarations with the one other API proposed so far that is able to start asynchronous tasks: `detach` and the `Task.Handle` that it returns.

First off, `detach` most of the time should not be used at all, because it does _not_ propagate task priority, task-local values or the execution context of the caller. Not only that but a detached task is inherently not _structured_ and thus may out-live its defining scope.

This immediately shows how `spawn let` and the general concept of child-tasks are superior to detached tasks. They automatically propagate all necessary information about scheduling and metadata necessary for execution tracing. And they can be allocated more efficiently than detached tasks.

> So while in theory one can think of `spawn let` as introducing a (hidden) `Task.Handle` or future, which is created at the point of declaration of the `spawn let` and whose value is retrieved at the `await` in practice, this comparison fails to notice the primary strenght of spawn lets: structured concurrency child-tasks.

Reviewer: `strenght` → `strength`

```suggestion
So while in theory one can think of `spawn let` as introducing a (hidden) `Task.Handle` or future, which is created at the point of declaration of the `spawn let` and whose value is retrieved at the `await` in practice, this comparison fails to notice the primary strength of `spawn let`: structured concurrency child-tasks.
```
